app/controllers/financialOverview/PaymentsAndChargeDetailsController.scala

```diff
@@ -0,0 +1,178 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.financialOverview
+
+import controllers.actions._
+import models.LocalDateBinder._
+import models.financialStatement.PaymentOrChargeType.{AccountingForTaxCharges, getPaymentOrChargeType}
+import models.financialStatement.SchemeFSChargeType.{PSS_AFT_RETURN, PSS_AFT_RETURN_INTEREST, PSS_OTC_AFT_RETURN}
+import models.financialStatement.{PaymentOrChargeType, SchemeFS}
+import models.requests.IdentifierRequest
+import models.{ChargeDetailsFilter, Submission}
+import play.api.Logger
+import play.api.i18n.{I18nSupport, Messages, MessagesApi}
+import play.api.libs.json.{JsObject, Json}
+import play.api.mvc._
+import renderer.Renderer
+import services.financialOverview.PaymentsAndChargesService
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import uk.gov.hmrc.viewmodels.{Html, NunjucksSupport}
+import utils.DateHelper.dateFormatterDMY
+
+import java.time.LocalDate
+import javax.inject.Inject
+import scala.concurrent.{ExecutionContext, Future}
+
+class PaymentsAndChargeDetailsController @Inject()(
+                                                    override val messagesApi: MessagesApi,
+                                                    identify: IdentifierAction,
+                                                    allowAccess: AllowAccessActionProviderForIdentifierRequest,
+                                                    val controllerComponents: MessagesControllerComponents,
+                                                    paymentsAndChargesService: PaymentsAndChargesService,
+                                                    renderer: Renderer
+                                                  )(implicit ec: ExecutionContext)
+  extends FrontendBaseController
+    with I18nSupport
+    with NunjucksSupport {
+
+  private val logger = Logger(classOf[PaymentsAndChargeDetailsController])
+
+  def onPageLoad(srn: String, pstr: String, period: String, index: String,
+                 paymentOrChargeType: PaymentOrChargeType, version: Option[Int],
+                 submittedDate: Option[String], journeyType: ChargeDetailsFilter): Action[AnyContent] =
+    (identify andThen allowAccess()).async {
+    implicit request =>
+      paymentsAndChargesService.getPaymentsForJourney(request.idOrException, srn, journeyType).flatMap { paymentsCache =>
+        val schemeFS: Seq[SchemeFS] = getFilteredPayments(paymentsCache.schemeFS, period, paymentOrChargeType)
+        buildPage(schemeFS, period, index, paymentsCache.schemeDetails.schemeName, srn, pstr, paymentOrChargeType, journeyType, submittedDate, version)
+      }
+  }
+
+  private def buildPage(
```

Comment:
Warnings in this file about too many parameters. Either exclude warnings via  // scalastyle:off parameter.number or find a way to reduce parameters.

Reply:
Done